### PR TITLE
QNTM-1228: Some nodes do not de-serialize from JSON correctly

### PR DESCRIPTION
**Pull request by ColinDayOrg**
### Purpose
This PR addresses an issue found with deserialization from JSON format where the order of hooking uop events and adding ports was incorrect. There are two main aspects to this PR; first altering the base NodeModel class so that the order of operations enables correct node functionality, and second removing any previous work-arounds that were added to hide the base issue.
 
### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [n/a] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [wip] The level of testing this PR includes is appropriate
- [n/a] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [n/a] Snapshot of UI changes, if any.
- [n/a] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @QilongTang 

### FYIs

@alfarok @jnealb

### CODE REVIEW

**Comment by ColinDayOrg** [comment]
Please see QNTM-1228 JIRA issue for details about testing methodology and results.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 1]
```diff
@@ -959,6 +959,11 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
             inputNodes = new Dictionary<int, Tuple<int, NodeModel>>();
             outputNodes = new Dictionary<int, HashSet<Tuple<int, NodeModel>>>();
 
+            // Initialize the port events
+            // Note; It is important that this occurs before the ports are added next
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: `Note:`?

**Comment by mjkkirschner** [comment]
@ColinDayOrg is this ready for review or is there more to do here? Like the race condition fix?

**Comment by ColinDayOrg** [comment]
The race condition issue is due to a regression from Dynamo 1.3, and has nothing to do with JSON (i.e it has to do with loading a file in automatic mode regardless of if the file is in XML or JSON format). The issue has been logged as QNTM-1392.

This PR (so far) includes a fix for the Unit Types node issue, but does not include the changes required to make at least one serialization test fail without these changes. After some investigation it appears that the JSON file that is saved in the serialization tests is somehow different from physically saving a JSON file from sandbox. I am currently trying to determine where the difference occurs, but it is possible that the investigation may be outside the desired scope of this specific task.

If it is decided that the test investigation is outside what is desired for this task then the PR should be good to go, assuming that the CI tests all pass. I am running CI tests now.

**Comment by mjkkirschner** [comment]
@ColinDayOrg well one big difference is the serialized files during testing have no view block.

**Comment by ColinDayOrg** [comment]
So far it seems that there is some different processing before saving that causes a difference when running the tests. The issue is that the in/out nodes count is different when loading a manually saved JSON file versus a JSON file saved during a test run.

**Comment by mjkkirschner** [comment]
@ColinDayOrg I'm fixing a bug in the inputs block right now (if thats what you're referring to) I did not deserialize them correctly.

**Comment by ColinDayOrg** [comment]
@mjkkirschner The JSOM loading  defect can be detected by comparing the Workspace.Nodes.InputNodes.Count or Workspace.Nodes.OutputNodes.Count between a loaded XML file and a loaded JSON file. Currently we only compare the Inports.Count and Outports.Count from XML and JSON files, so the issue was not detected. 

However, after adding the additional check to the test framework to check the InputNode and OutputNode count I found that the tests were still passing. Manually loading the XML and saving/reloading the JSON I can definitively see the difference.

**Comment by mjkkirschner** [comment]
🚢  it!

**Pull request merged**